### PR TITLE
Overhaul and re-theme documentation

### LINKED ISSUE #1941: Switch to dirhtml builder for cleaner Sphinx/RTD URLs
- Use `dirhtml` builder instead of `html` locally and on RTD for cleaner URLs: 
  - Documentation: https://www.sphinx-doc.org/en/master/usage/builders/index.html
  - Before: `https://urllib3.readthedocs.io/en/latest/user-guide.html`
  - After: `https://urllib3.readthedocs.io/en/latest/user-guide` :sparkles: 
- Use RTD redirects of `.html` pages to their non-`.html` counterparts to preserve links: https://blog.readthedocs.com/user-defined-redirects/

### PR DESCRIPTION
**Pull request by sethmlarson**
- Switches the documentation to a new theme, [Furo](https://github.com/pradyunsg/furo), made by @pradyunsg
  - ~The only place I think we need is the urllib3 icon in the brand section~ (done)
- Changes some headings while maintaining the previous fragments 
- Updated many `code-block`s to use the proper `pycon` (Python Console) instead of assuming `python` as default
- Created a favicon for our documentation
- Split the Sponsors section out of Contributing (this will be updated more in the future)
- Fixed many directives from being overspaced (like `.. note::`, etc) so they don't appear doubly indented or quoted in the documentation.
- Favor "TLS" over "SSL" in prose

I tried to go through every section of our documentation and look for broken things or places that looked off. Would be good to make sure I didn't miss anything but always can be corrected in the future.

Some todos for myself:
- I'd like to have prose at the beginning of each section in the API reference
- ~The top-level APIs section might need some pruning with links to the deeper documentation.~ (done)
 Might be good to update the sections to only mention the sub-module they're defined in if they're not
 within the `urllib3` namespace.
- ~Search with `dirhtml` mode doesn't seem to be working well.~ (fixed in Furo `master`)

### CODE REVIEW

**Comment by pradyunsg** [comment]
> The only place I think we need is the urllib3 icon in the brand section

Okie. I'll make this possible in... *checks clock* 3 hours after dinner. 🙃

**Comment by sethmlarson** [comment]
@pradyunsg No rush! We've lived without our fancy banner plastered on everything for so long we'll survive without it for some time ;)

**Review by pquentin — APPROVED** [review]
This is awesome, thanks! Note that you can enable RTD autobuild on pull request: https://docs.readthedocs.io/en/stable/guides/autobuild-docs-for-pull-requests.html. It makes it even easier to review docs change.

Do we really want to switch to `dirhtml`? This seems to be a lot of work and a lot of churn for an arguably small benefit. Cool URIs don't change, after all

**Comment by sethmlarson** [comment]
I can revert the `dirhtml` change for now, my end-goal is to have our documentation potentially hosted under `urllib3.dev` and URLs would look better as `urllib3.dev/user-guide` instead of `urllib3.dev/user-guide.html`. I guess personal preference?

**Comment by pquentin** [comment]
If we're going to switch to another domain entirely anyway, then yeah URLs without `.html` *do* look nicer and we can switch. But maybe we should do that only on urllib3.dev?

**Comment by pradyunsg** [comment]
Whoops. New release w/ support for logos (and a few more tweaks) is up. :)

**Comment by sethmlarson** [comment]
@pquentin Okay I've reverted back to the HTML builder, got the `html_logo` working with a tiny bit of custom CSS to hide the short doc description under the logo.

**Comment by sethmlarson** [comment]
We should have RTD builds active from now on!

**Comment by pradyunsg** [comment]
I don't see it! :o

**Review by pquentin — APPROVED** [review]
@sethmlarson This is awesome, thank you! Those docs look really good, it was well worth the effort.

Thank you @pradyunsg :)

**Comment by sethmlarson** [comment]
Yeah RTD failed on https://github.com/pradyunsg/furo/issues/5 which is fixed in master. I'll cycle this PR once a new release is out to make sure things work before we go live :) Just another reason to use RTD per-PR builds.

**Comment by pradyunsg** [comment]
Release is out. :)

File: docs/sponsors.rst [hunk 1]
```diff
@@ -0,0 +1,33 @@
+Sponsors
```
Comment (pradyunsg) [inline docs/sponsors.rst:1]:
```suggestion
:hide-toc:

Sponsors
```

This'll look nicer.

**Comment by sethmlarson** [comment]
Huzzah! Commit status is now showing up, just had to grant access to the RTD OAuth Application to the urllib3 org.

File: docs/conf.py [hunk 2]
```diff
@@ -48,227 +33,50 @@ def __getattr__(cls, name):
...
 
 # The name of the Pygments (syntax highlighting) style to use.
-pygments_style = 'sphinx'
-
-# A list of ignored prefixes for module index sorting.
-#modindex_common_prefix = []
-
-
-# -- Options for HTML output ---------------------------------------------------
+pygments_style = "friendly"
```
Comment (pradyunsg) [inline docs/conf.py:71]: OOoooo. I *really* should find a good theme to use by default in furo itself. :)

**Comment by sethmlarson** [comment]
@pradyunsg Looks like there are still issues with `lru_cache` and Python 3.7: https://readthedocs.org/projects/urllib3/builds/11836919/

File: noxfile.py [hunk 3]
```diff
@@ -104,10 +112,8 @@ def lint(session):
     session.run("flake8", "--version")
     session.run("black", "--version")
     session.run("mypy", "--version")
-    session.run(
-        "black", "--check", "src", "dummyserver", "test", "noxfile.py", "setup.py"
-    )
-    session.run("flake8", "setup.py", "docs", "dummyserver", "src", "test")
+    session.run("black", "--check", *SOURCE_FILES)
+    session.run("flake8", *SOURCE_FILES)
```
Comment (pradyunsg) [inline noxfile.py:116]: nit: consider adopting <https://pre-commit.com>. :)

**Review by pradyunsg — COMMENTED** [review]
This looks wonderful! YAY! \o/

(aside: I usually make a few comments prefixed with "nit: " in my code reviews. These are self-acknowledged nit-picky comments. I'd encourage you to ignore those if you disagree with them, since I don't think they're worthwhile investment of time/energy to discuss them.)

**Comment by pradyunsg** [comment]
Ooops. We missed one!

**Comment by pradyunsg** [comment]
That should fix it.

**Comment by pradyunsg** [comment]
https://urllib3--1950.org.readthedocs.build/en/1950/

IT'S ALIVE!

**Comment by sethmlarson** [comment]
Let's goooooo :shipit:

**Comment by sethmlarson** [comment]
Big thank you again to @pradyunsg for the theme and all this interaction and support you've provided us :bowing_man: :heart:

**Comment by sethmlarson** [comment]
Documentation is live: https://urllib3.readthedocs.io

**Comment by pradyunsg** [comment]
Thanks for adopting a theme that's not even fully documented yet. :)

File: docs/sponsors.rst (hunk 1 above)
Comment (pradyunsg) [reply docs/sponsors.rst:1]: @sethmlarson *nudge*

Comment (sethmlarson) [reply docs/sponsors.rst:1]: Thank you for the nudge! I need to implement this in my sponsor + opencollective branch.

**Pull request merged**